frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/FinallyNode.tsx

```diff
@@ -0,0 +1,118 @@
+import { PipelineVisualizationTask } from '../detail-page-tabs/pipeline-details/PipelineVisualizationTask';
+import {
+  BUILDER_NODE_ADD_RADIUS,
+  NODE_WIDTH,
+  NODE_HEIGHT,
+  FINALLY_NODE_PADDING,
+  FINALLY_NODE_VERTICAL_SPACING,
+  FINALLY_ADD_LINK_TEXT_HEIGHT,
+  BUILDER_NODE_ERROR_RADIUS,
+  FINALLY_ADD_LINK_SIZE,
+} from './const';
+import PlusNodeDecorator from './PlusNodeDecorator';
+import ErrorNodeDecorator from './ErrorNodeDecorator';
+import TaskList from './TaskList';
+import { getFinallyTaskHeight } from './utils';
+
+const FinallyNode: React.FC<{ element: Node }> = ({ element }) => {
+  const {
+    clusterTaskList = [],
+    namespaceTaskList = [],
+    task,
+    pipeline,
+    pipelineRun,
+    disableBuilder,
+  } = element.getData();
+  const { t } = useTranslation();
+  const { finallyTasks = [], finallyListTasks = [] } = task;
+  const allTasksLength = finallyTasks.length + finallyListTasks.length;
+  return (
+    <>
+      <rect
+        fill="transparent"
+        strokeWidth={1}
+        stroke="var(--pf-global--BorderColor--light-100)"
+        width={NODE_WIDTH + FINALLY_NODE_PADDING * 2}
+        height={getFinallyTaskHeight(allTasksLength, disableBuilder)}
+        rx="20"
+        ry="20"
+      />
+
+      {finallyTasks.map((ft, i) => (
+        <g
+          key={ft.name}
+          transform={`translate(${FINALLY_NODE_PADDING}, ${NODE_HEIGHT * i +
+            FINALLY_NODE_VERTICAL_SPACING * i +
+            FINALLY_NODE_PADDING})`}
+          onClick={ft.onTaskSelection && !disableBuilder ? () => ft.onTaskSelection() : null}
+        >
+          <PipelineVisualizationTask
+            pipelineRunName={pipelineRun?.metadata?.name}
+            task={ft}
+            pipelineRunStatus={pipelineRun && pipelineRunFilterReducer(pipelineRun)}
+            namespace={pipeline?.metadata?.namespace}
+            disableTooltip={!disableBuilder}
+            selected={ft.selected}
+            width={NODE_WIDTH}
+            height={NODE_HEIGHT}
+          />
+          {ft.error && !disableBuilder && (
+            <ErrorNodeDecorator
+              x={BUILDER_NODE_ADD_RADIUS / 2}
+              y={BUILDER_NODE_ERROR_RADIUS / 4}
+              errorStr={ft.error}
+            />
+          )}
+        </g>
+      ))}
+      {!disableBuilder &&
+        finallyListTasks.map((flt, i) => (
+          <g
+            key={flt.name}
+            transform={`translate(${FINALLY_NODE_PADDING}, 
+              ${NODE_HEIGHT * (i + finallyTasks.length) +
+                FINALLY_NODE_VERTICAL_SPACING * (i + finallyTasks.length) +
+                FINALLY_NODE_PADDING})`}
+          >
+            <TaskList
+              width={NODE_WIDTH}
+              height={NODE_HEIGHT}
+              listOptions={[...clusterTaskList, ...namespaceTaskList]}
+              onRemoveTask={flt.onRemoveTask}
+              onNewTask={flt.convertList}
+            />
+          </g>
+        ))}
+      {!disableBuilder && (
```

Comment:
What do you think about removing `disableBuilder` prop and making two components with reusable sections from both. I'm not sure you're saving maintenance effort here, you seem to just condition a bunch of stuff and your rendering logic is almost all within the return statement, so it would be easy to break up into two components. `BuilderFinallyNode` and `FinallyNode`

Reply:
Separated into two different components.